specifyweb/frontend/js_src/lib/components/DataModel/uniquenessRules.ts

```diff
@@ -0,0 +1,184 @@
+import React from 'react';
+
+import { useCachedState } from '../../hooks/useCachedState';
+import { formsText } from '../../localization/forms';
+import { ajax } from '../../utils/ajax';
+import { getCache, setCache } from '../../utils/cache';
+import type { GetOrSet, RA } from '../../utils/types';
+import { formatConjunction } from '../Atoms/Internationalization';
+import { load } from '../InitialContext';
+import type { SerializedResource } from './helperTypes';
+import type { LiteralField, Relationship } from './specifyField';
+import type { SpLocaleContainerItem, Tables } from './types';
+
+export type UniquenessRules = {
+  readonly [TABLE in keyof Tables]?: RA<{
+    readonly id: number | null;
+    readonly fields: RA<SerializedResource<SpLocaleContainerItem>>;
+    readonly scope: SerializedResource<SpLocaleContainerItem> | null;
+    readonly isDatabaseConstraint: boolean;
+
+    // This property is assigned on the frontend and is not saved to the backend
+    readonly uniqueId?: number;
+  }>;
+};
+
+export type UniquenessRule = Exclude<
+  UniquenessRules[keyof Tables],
+  undefined
+>[number];
+
+const setInitialRules = async (): Promise<UniquenessRules> =>
+  import('./schema')
+    .then(async ({ fetchContext }) => fetchContext)
+    .then(async (schema) =>
+      load<UniquenessRules>(
+        `/businessrules/uniqueness_rules/${schema.domainLevelIds.discipline}/`,
+        'application/json'
+      )
+    )
+    .then((data) => {
+      setCache(
+        'businessRules',
+        'uniqueRules',
+        data as Readonly<Record<string, string>>
+      );
+      return data;
+    });
+
+export const fetchContext = setInitialRules();
+
+export function getUniquenessRules(): UniquenessRules | undefined;
+export function getUniquenessRules<TABLE_NAME extends keyof Tables>(
+  model: TABLE_NAME
+): UniquenessRules[TABLE_NAME] | undefined;
+export function getUniquenessRules<TABLE_NAME extends keyof Tables>(
+  model?: TABLE_NAME
+): UniquenessRules | UniquenessRules[TABLE_NAME] {
+  const uniquenessRules = getCache('businessRules', 'uniqueRules');
+  return uniquenessRules === undefined
+    ? undefined
+    : model === undefined
+    ? uniquenessRules
+    : uniquenessRules[model.toLowerCase() as keyof Tables];
+}
+
+export function useModelUniquenessRules<TABLE extends keyof Tables>(
+  modelName: TABLE
+): readonly [
+  ...modelRules: GetOrSet<UniquenessRules[TABLE]>,
```

Comment:
Yes, it can be simplified. 
I modeled the `UniquenessRules` type from the returned JSON structure of the endpoint. 
Though simplifying the type would be warranted as I am unsure if we would ever need to explicitly specify or use the entire endpoint value besides the initial `load` call

Reply:
Fixed in 6f1043eed4a14114cb501aaf721df64bd7d0dcdc